### PR TITLE
Assignment6 -Spencer West & Matthew Powers

### PR DESCRIPTION
**Pull request by SaaWest**
@SaaWest @Matthew-Powers

### CODE REVIEW

**Comment by Joshua-Lester3** [comment]
I'll review this :)

File: CalculateTests/ProgramTests.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿using Xunit;
+using Calculate;
+
+namespace CalculateTests;
+
+public class ProgramTests
```
Comment (Joshua-Lester3) [inline CalculateTests/ProgramTests.cs:6]: Input and output of Program.Main is not tested. The properties are tested, but not the application that you can run on the command line

**Review by Joshua-Lester3 — COMMENTED** [review]
- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ❌ I think the single quotes mean that's what they should be identified as. 
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌ No test for Reader (ReadLine) property
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ❌ Not implemented as a property. Also it's static and I believe properties have to be instance members.
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ❌ Everything is good here except for the 'integers' part. I would argue the current implementation as doubles is fine, however.
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ❌ The pass/fail on this requirement could change depending on how strictly one follows "entirely". My interpretation is you have to use the .Split(), pattern matching, and logical and operators, but that could be wrong. If I'm right, my suggestion is to create your own method for parsing the string instead of using .TryParse().
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌ Not instantiated because it appears that calculator only has static methods/fields
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase` ❌ not implemented
- Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects ❌ not implemented
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ❌ not implemented

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ❌ generic version isn't needed with xunit, but the Equal assertion is not used
- All of the above should be unit tested ❌ Program.Main is not tested
- Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+public class Program
+{
+    public Func<string?> Reader { get; init; } = Console.ReadLine;
+    public Action<string> Writer { get; init; } = Console.WriteLine;
+
+    public Program()
+    {
+    }
+
+    public static void Main()
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Calculate/Program.cs:12]: implemented changes

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿namespace Calculate;
+
+public class Calculator
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Calculate/Calculator.cs:3]: made the change thank you for bringing it to my attention

File: CalculateTests/ProgramTests.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        string line = "Can you Write this?";
+        string output = "";
+
+        Program program = new() { Writer = (writer) => output = writer };
+        program.Writer(line);
+        Assert.NotEqual("Writer writing", output);
+    }
+
+    [Fact]
+    public void Program_CreateNewProgram_Success()
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply CalculateTests/ProgramTests.cs:31]: will remove I've gotten it of reading the instructions and implement a tests then writing the code

**Comment by EthanFlorko** [comment]
I'll review this

File: CalculateTests/CalculatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    [Theory]
+    [InlineData("2 # 2")]
+    public void Calculator_TryCalculateInvalidOperation_TestFailure(string expression)
+    {
+        Calculator calculator = new();
+        var tests = calculator.TryCalculate(expression, out _);
+        Assert.False(tests);
+    }
+}
```
Comment (EthanFlorko) [inline CalculateTests/CalculatorTests.cs:107]: Testing covers multiple scenarios and minimal duplication. Checking for both invalid and valid results. Handling edge cases and error conditions as well.

**Comment by EthanFlorko** [comment]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ❌
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ❌
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ❌
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ❌
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: CalculateTests/CalculatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+public class CalculatorTests
+{
+    [Theory]
+    [InlineData(1, 1)]
+    public void Calculator_Add_TestSuccess(int x, int y) 
+    {
+        var z = Calculator.Add(x, y);
+        
+        Assert.Equal(2, z);
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply CalculateTests/CalculatorTests.cs:14]: thank you for the explanation and clarity

File: CalculateTests/CalculatorTests.cs (hunk 5 above)
Comment (SaaWest) [reply CalculateTests/CalculatorTests.cs:107]: will implement more scenarios in tests

File: CalculateTests/ProgramTests.cs (hunk 1 above)
Comment (SaaWest) [reply CalculateTests/ProgramTests.cs:6]: tried to test main with intellitect tools but it could not resolve issues despite the tool being declared in the csproj file.

File: CalculateTests/ProgramTests.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+﻿using Xunit;
+using Calculate;
+
+namespace CalculateTests;
+
+public class ProgramTests
+{
+    [Fact]
+    public void Program_ReadWrite_ValidTestSuccess()
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply CalculateTests/ProgramTests.cs:9]: added

File: CalculateTests/CalculatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        Calculator calculator = new();
+        Assert.NotNull(calculator);
+    }
+
+    [Theory]
+    [InlineData(1, 1)]
+    public void Calculator_Add_TestSuccess(double x, double y) 
+    {
+        var z = Calculator.Add(x, y);
+        Assert.True(z.Equals(2));
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply CalculateTests/CalculatorTests.cs]: made changes

File: Calculate/Calculator.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        //char operation = input[1][0];
+        if (!double.TryParse(input[0], out double operand1) || !double.TryParse(input[2], out double operand2)) 
+        { 
+            return false; 
+        }
+        if (!MathematicalOperations.ContainsKey(input[1][0]) ) 
+        { 
+            return false; 
+        }
+        var op = MathematicalOperations[input[1][0]];
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Calculate/Calculator.cs]: made recommended changes

**Comment by SaaWest** [comment]
disclaimer the ProgramBase isn't implemented as I had problems with csproj reading the file path

File: Calculate/Calculator.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+using System.Numerics;
+
+namespace Calculate;
+
+public class Calculator<T> where T : INumber<T>
+{
+    public Calculator()
+    {
+    }
+#pragma warning disable CA1000
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:11]: Justificication for warning disable

File: Calculate/Calculator.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
+﻿using System.Globalization;
+using System.Numerics;
+
+namespace Calculate;
+
+public class Calculator<T> where T : INumber<T>
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:6]: Consider taking in two type parameters in order to return a different precision than what is equated with (ex: int / int = double)

File: CalculateTests/CalculatorTests.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    public void Calculator_TryCalculateMultiplication_TestSuccess(string expression, int output)
+    {
+        Calculator<int> calculator = new();
+        calculator.TryCalculate(expression, out int expected);
+
+        Assert.Equal(output, expected);
+    }
+
+    [Theory]
+    [InlineData("10 / 2", 5)]
```
Comment (BenjaminMichaelis) [inline CalculateTests/CalculatorTests.cs:76]: What if you do 10/6?
